enhancements/microshift/multus-cni-for-microshift.md

```diff
@@ -0,0 +1,445 @@
+is a connectivity.
+
+Tests for other networking plugins will be designed and implemented when plugins are planned for support.
+
+### Graduation Criteria
+
+Multus CNI for MicroShift is targeted to be GA next release.
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+N/A
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+Because both Multus and MicroShift RPMs will be built from the same spec file, they will share the
+same version and it is expected that they are updated together following MicroShift upgrade rules
+depending on type of operating system (ostree-based or regular RPM).
+
+Considering only the manifests, we know that on each start MicroShift will apply manifests forcefully
+overwriting any differences. However, MicroShift does not have any uninstall capabilities.
+If manifests ever change, for example some ConfigMap is renamed, then these old parts will
+keep existing in the database. This could have undesirable consequences of having two Multus DaemonSets
+if the name or the namespace of original DaemonSet changes. To make the transition to thick Multus
+smooth, we should not deviate from already existing resource names present in OpenShift's Multus
+manifests. This problem is not Multus specific - it is how MicroShift works and it is not part of this
+enhancement addressing this shortcoming.
+
+Because of the way Multus works (checks Annotations when needed, executed when networking must be
+set up, does not keep its own database beside cache) we can consider it to be mostly stateless.
+When Multus CNI binary is executed by kubelet/CRI-O it will setup Pod's network according to the
+configs and Annotations (not a database with "expected states"), therefore updating Multus does not
+require migration strategies.
+
+If the schema of NetworkAttachmentDefinitions CRD changes, instead of deploying a mutating webhook
+(which would use resources idling but actually perform actions on first start), we can just suggest
+users to update their CRs.
+
+### Version Skew Strategy
+
+Building Multus and MicroShift RPMs from the same spec file means Multus should be updated together
+with MicroShift which means there should not be any version skew between MicroShift and Multus.
+This might change with introduction of multi-node deployments of MicroShift.
+
+### Operational Aspects of API Extensions
+
+N/A
+
+#### Failure Modes
+
+If Multus (or any delegate CNIs it executes) fails, a new Pod will be stuck in "ContainerCreating" status
+and none of the Pod's containers will start. This can happen if the CNI configuration provided in
+NetworkAttachmentDefinition is incorrect or when Pod's Annotation contains NAD that does not exist.
+In such cases, user needs to verify its manifests.
+
+Pods without Multus' Annotation will be set up with the default CNI (ovn-kubernetes) and should not
+have increased CNI failure rate.
+
+#### Support Procedures
+
```

Comment:
One other support procedure might be to configure multus to redirect log to local file.
Multus thin plugin mode doesn't provide stdout logs in its container like other long-running CNI (ovn-kubernetes) does. If any failure in multus binary itself, the stdout can only be retrieved by updating multus cni config to redirect log to local file: https://github.com/openshift/multus-cni/blob/master/docs/configuration.md#writing-to-a-log-file

Reply:
Multus messages in pod events are mainly for users (just for 'events'). More multus messages are captured by crio/kubelet journal logs, so please describe that in this section.

Reply:
Example

```
[fedora@kube-node-1 ~]$ journalctl -u kubelet | tail
(snip)
Feb 06 14:27:31 kube-node-1 kubelet[13551]: E0206 14:27:31.893101   13551 kuberuntime_manager.go:1172] "CreatePodSandbox for pod failed" err="rpc error: code = Unknown desc = failed to create pod network sandbox k8s_fedora_default_228fdcfa-dfc3-46de-b0b5-d996457f7d7e_0(dae451e8aaf3e9bff94ce8c1e2dc3918b46db2334f88f028c5e5d59b8156bf06): error adding pod default_fedora to CNI network \"multus-cni-network\": plugin type=\"multus\" name=\"multus-cni-network\" failed (add): [default/fedora/228fdcfa-dfc3-46de-b0b5-d996457f7d7e:macvlan-conf-1]: error adding container to network \"macvlan-conf-1\": Link not found" pod="default/fedora"
Feb 06 14:27:31 kube-node-1 kubelet[13551]: E0206 14:27:31.893286   13551 pod_workers.go:1298] "Error syncing pod, skipping" err="failed to \"CreatePodSandbox\" for \"fedora_default(228fdcfa-dfc3-46de-b0b5-d996457f7d7e)\" with CreatePodSandboxError: \"Failed to create sandbox for pod \\\"fedora_default(228fdcfa-dfc3-46de-b0b5-d996457f7d7e)\\\": rpc error: code = Unknown desc = failed to create pod network sandbox k8s_fedora_default_228fdcfa-dfc3-46de-b0b5-d996457f7d7e_0(dae451e8aaf3e9bff94ce8c1e2dc3918b46db2334f88f028c5e5d59b8156bf06): error adding pod default_fedora to CNI network \\\"multus-cni-network\\\": plugin type=\\\"multus\\\" name=\\\"multus-cni-network\\\" failed (add): [default/fedora/228fdcfa-dfc3-46de-b0b5-d996457f7d7e:macvlan-conf-1]: error adding container to network \\\"macvlan-conf-1\\\": Link not found\"" pod="default/fedora" podUID="228fdcfa-dfc3-46de-b0b5-d996457f7d7e"

[fedora@kube-node-1 ~]$ journalctl -u crio | tail
(snip)
Feb 06 14:28:44 kube-node-1 crio[11778]: 2024-02-06T14:28:44+09:00 [verbose] Del: default:fedora:228fdcfa-dfc3-46de-b0b5-d996457f7d7e:cbr0:eth0 {"cniVersion":"0.3.1","name":"cbr0","plugins":[{"delegate":{"hairpinMode":true,"isDefaultGateway":true},"type":"flannel"},{"capabilities":{"portMappings":true},"type":"portmap"}]}
Feb 06 14:28:44 kube-node-1 crio[11778]: 2024-02-06T14:28:44+09:00 [debug] conflistDel: &{1ae5a8f8270afbe7faa5e54ed636692bae72fb73ab9ca0132e8e66307b34b696 /var/run/netns/bc624229-b577-40e3-ad41-7af830a909bc eth0 [[IgnoreUnknown true] [K8S_POD_NAMESPACE default] [K8S_POD_NAME fedora] [K8S_POD_INFRA_CONTAINER_ID 1ae5a8f8270afbe7faa5e54ed636692bae72fb73ab9ca0132e8e66307b34b696] [K8S_POD_UID 228fdcfa-dfc3-46de-b0b5-d996457f7d7e] [IgnoreUnknown 1] [K8S_POD_NAMESPACE default] [K8S_POD_NAME fedora] [K8S_POD_INFRA_CONTAINER_ID 1ae5a8f8270afbe7faa5e54ed636692bae72fb73ab9ca0132e8e66307b34b696] [K8S_POD_UID 228fdcfa-dfc3-46de-b0b5-d996457f7d7e]] map[] }, {"cniVersion":"0.3.1","name":"cbr0","plugins":[{"delegate":{"hairpinMode":true,"isDefaultGateway":true},"type":"flannel"},{"capabilities":{"portMappings":true},"type":"portmap"}]}
```

Reply:
Thanks, added kubelet, but I don't see these logs in crio even though I configured `log_level = "trace"`.
I think it might be worth it after all to enable logging in the multus itself (`--multus-log-file`, `--multus-log-level`), WDYT?

Reply:
- as far as I checked upstream, crio captures the logs as above. I don't know why. Does microshift uses special crio?
- it is trade-off to disk space and debuggability. I don't have clear picture how to troubleshoot in microshift. If you have such troubleshoot process, it might be added.

Reply:
Added `log-file` as open question.
I don't know about the crio - tried setting the log level, playing with different log settings for multus and still didn't get that kind of log. Maybe I'll find something during implementation phase